### DISCUSSION TITLE
`MSBUILD : error MSB1001: Unknown switch` with `tauri dev` and `tauri build`

### DISCUSSION THREAD

**Discussion by lorrding** [tauri-apps/tauri#9046] [Q&A] [answered]
Hi,
I'm having issues running the quick start guide from the Tauri doc.
I'm on a corporate laptop with group policies so I **cannot** run any *"untrusted"*  .exe files outside of the `my-app` folder in my user profile (`C:\Users\%USER%\MyApp`). I have Rust, Cargo and Rustup installed inside this folder and added to my custom path profile.

When running `tauri info` I get:
```
> tauri-app@0.0.0 tauri
> tauri info


[✘] Environment
    - OS: Windows 10.0.19045 X64
    ✔ WebView2: 122.0.2365.59
    ✘ Couldn't detect any Visual Studio or VS Build Tools instance with MSVC and SDK components. Download from https://aka.ms/vs/17/release/vs_BuildTools.exe
    ✔ rustc: 1.76.0 (07dca489a 2024-02-04)
    ✔ cargo: 1.76.0 (c84b36747 2024-01-18)
    ✔ rustup: 1.26.0 (5af9b9484 2023-04-05)
    ✔ Rust toolchain: windows-msvc (default)
    - node: 20.9.0
    - npm: 10.1.0

[-] Packages
    - tauri [RUST]: 1.6.0
    - tauri-build [RUST]: 1.5.1
    - wry [RUST]: 0.24.7
    - tao [RUST]: 0.16.7
    - @tauri-apps/api : not installed!
    - @tauri-apps/cli [NPM]: 1.5.7 (outdated, latest: 1.5.10)

[-] App
    - build-type: bundle
    - CSP: unset
    - distDir: ../ui
    - devPath: ../ui
```
The rust toolchains are from the standalone msvc version of rust 1.76.0 and linked using `erustup toolchain link [name] [path]`

As you can see, Tauri cannot seem to find the VS Build Tools even though it is installed and working. MSVC is part of the trusted softwares of my company, so it's installed like on any *normal* computer (in `C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\MSBuild\Current\Bin`)

What bugs me out is that when running `tauri dev`, tauri seems to call cargo which tries to build the project using MSVC. But MSVC exit with an error:
```
> tauri-app@0.0.0 tauri
> tauri dev

        Info Watching C:\Users\%USER%\MyApp\code\TestTauri\src-tauri for changes...
Microsoft (R) Build Engine version 16.9.2+58c36d143 for .NET Framework
Copyright (C) Microsoft Corporation. All right reserved.

MSBUILD : error MSB1001: Unknown switch.
Switch: --no-default-features
```

Any idea how I can either tell cargo to not use this "--no-default-features" flag, find why cargo use a flag that msvc does not know or tell Tauri where the build tools are ?

Thanks for your help!

EDIT:
Digging a little deeper into Tauri/cli I found out that `tauri info` will run a file named `vswhere.exe`, but I can find it inside the node_modules and assumed it is compiled inside `cli.win32-x64-msvc.node`. I downloaded the .exe and tried executing the arguments provided in the info function:
https://github.com/tauri-apps/tauri/blob/a77be9747443ffc29c34160b55893483bb5f0d74/tooling/cli/src/info/env_system.rs#L35-L49):
```ps
c:\Users\%USER%\MyApp\app\vswhere.exe -prerelease -products * -requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 -requires Microsoft.VisualStudio.Component.Windows10SDK.* -format json -utf8
```
I get a json output with a lot of information. I don't know if it's a good idea to share every detail but what I understand is that VSWhere.exe gave me something back. In the Tauri dev function, I can see that it *"only"* checks if the output is an empty string.
See:
https://github.com/tauri-apps/tauri/blob/a77be9747443ffc29c34160b55893483bb5f0d74/tooling/cli/src/info/env_system.rs#L205-L214
So in theory it ***should*** be able to find the build tools but can't. I'm stuck again and don't understand why.

I still need help or advice if anyone as an idea!

**Comment by FabianLars** [comment]
Hi, does it work if you cd into the src-tauri dir and run `cargo build --no-default-features` ? That's essentially what tauri does and if it does not work it may be a question for the rust devs/community. (ignoring the vswhere issues for now)

**Reply by lorrding** [reply]
Hi, thank you for your answer.
I just tried running ``cargo build --no-default-features`` inside src-tauri and it worked just fine. But again running outside with npm run tauri dev exit with the same error as before.

Also worth noting: this bug seems to be fixed in V2 Beta. I tried running the sample project from tauri beta with ``tauri dev`` and it works perfectly! The issue is that I'm not sure if using the beta right now is a good idea as it is well ... a beta.

Any idea what could be causing v1 to fail ? Is there any flag or setting inside tauri.conf I could set to tinker with the way tauri runs cargo ? If I understand correctly, tauri seems to pass the `--no-default-feature` to MSBuild instead of Cargo and that's what's causing the error. I didn't find anything on the CLI doc or when digging in the source code.

**Reply by FabianLars** [reply]
Tauris never invokes msbuild itself. That's what's confusing me. The --no-default-features flag is given to **cargo** to disable the custom-protocol feature in dev. The feature flag was removed in the latest beta hence we didn't need to pass the flag.

Somehow the custom cargo or whatever must have changed how it consumes cli flags or something 🤔

**Reply by lorrding** [reply, marked as answer]
Update: I tried reinstalling the Build Tool to see if maybe there was an error in the installation. And I also tried creating a brand new project from scratch (I was building with the same project up until now). It now seems to dev and build just fine!

I think what was causing the error may be due to some link deep inside the conf not updating after installing/reinstalling the build tools or updating some dependencies with npm or cargo? 🤔 

Right now in a new project I don't have the issue, but I still have it in the previous one. I'll assume it's a mistake on my part while trying to follow the quick start guide.

Still, It would be great to have a little bit more control over the verbosity of cargo and the build tools when using --verbose on tauri. I tried it before posting here, but it was not very helpful and I found it hard to understand what Tauri is doing under the hood. Maybe something that could be documented or better referenced in v2?

Anyway, thank you for your time!

**Reply by FabianLars** [reply]
we'll keep that in mind, thanks for the feedback :)
